api/v3/auth.go

```diff
@@ -0,0 +1,612 @@
+		res.R(w, r, res.ErrBadRequest("parameters user, token cannot be empty"))
+		return
+	}
+
+	token, err := jwt.Parse(tokenStr, func(token *jwt.Token) (interface{}, error) {
+		if method, ok := token.Method.(*jwt.SigningMethodHMAC); !ok {
+			return nil, fmt.Errorf("unable to validate signing method: %v", token.Header["alg"])
+		} else if method != a.jwt.SigningAlgo {
+			return nil, errors.New("expect hs512 signing method")
+		}
+		return []byte(a.jwt.Key), nil
+	})
+	if err != nil {
+		res.R(w, r, res.ErrUnauthorized("invalid token", "error", err))
+		return
+	}
+	if !token.Valid {
+		res.R(w, r, res.ErrUnauthorized("invalid token"))
+		return
+	}
+
+	claims, ok := token.Claims.(jwt.MapClaims)
+	if !ok {
+		res.R(w, r, res.ErrBadRequest("invalid token claims"))
+		return
+	}
+	if v, ok := claims[claimUser].(string); !ok || v != user {
+		res.R(w, r, res.ErrBadRequest("user in token does not match request"))
+		return
+	}
+
+	// check expiry
+	if err := claims.Valid(); err != nil {
+		res.R(w, r, res.ErrBadRequest("invalid claims",
+			"error", err))
+		return
+	}
+
+	u, err := a.users.FindByUserName(user)
+	if err != nil {
+		res.R(w, r, res.ErrNotFound("user not found",
+			"user", user))
+		return
+	}
+	challenge, ok := claims[claimChallenge].(string)
+	if !ok || challenge != u.EmailVerificationToken {
+		res.R(w, r, res.ErrBadRequest("challenge in token is incorrect"))
+		return
+	}
+	if _, err := a.users.ValidateEmailVerificationToken(user, challenge); err != nil {
+		l.Errorw("unexpected error when validating user",
+			"error", err)
+		res.R(w, r, res.ErrInternalServer("unable to validate user", err))
+		return
+	}
+
+	l.Info("user verified")
+	res.R(w, r, res.MsgOK("user verified"))
+}
+
+// newAuthInterceptors creates unary and stream interceptors that validate
+// requests, for use with gRPC servers
+func (a *AuthService) newAuthInterceptors(exceptions ...string) (
+	unaryInterceptor grpc.UnaryServerInterceptor,
+	streamInterceptor grpc.StreamServerInterceptor,
+) {
+	exclude := make(map[string]bool)
+	for _, e := range exceptions {
+		exclude[e] = true
+	}
+
+	// unaryInterceptor handles all incoming unary RPC requests
+	unaryInterceptor = func(
+		ctx context.Context,
+		req interface{},
+		info *grpc.UnaryServerInfo,
+		handler grpc.UnaryHandler,
+	) (r interface{}, err error) {
+		if v, found := exclude[info.FullMethod]; !v && !found {
+			a.l.Debugw("requested RPC is an exception - skipping authentication",
+				"method", info.FullMethod)
+			if ctx, err = a.validate(ctx); err != nil {
+				return
+			}
+		}
+		if handler != nil {
+			return handler(ctx, req)
+		}
+		return
+	}
+
+	// streamInterceptor handles all incoming stream RPC requests
+	streamInterceptor = func(
+		srv interface{},
+		stream grpc.ServerStream,
+		info *grpc.StreamServerInfo,
+		handler grpc.StreamHandler,
+	) (err error) {
+		if v, found := exclude[info.FullMethod]; !v && !found {
+			a.l.Debugw("requested RPC is an exception - skipping authentication",
```

Comment:
What is this doing, and to what does it apply to? looks like a good place to add a quick explainer comment

Reply:
8b80cc4 - please let me know if this doens't clarify it! The tests also demonstrate what is going on:

https://sourcegraph.com/github.com/RTradeLtd/Temporal@api/v3/prototype/-/blob/api/v3/auth_test.go#L913:57

Reply:
Looks great!